Add role-mgr, to deploy ceph-mgr instances

tserong opened the pull request:
PROBLEMS:
- Error EINVAL: entity mgr.ses4-2 exists but key does not match
(when run over an existing cluster with auto-deployed mgrs)

TODO:
- remove
- rescind
- only mandatory on Luminous+

ALSO:
- add mgr support to engulfer

Fixes: https://github.com/SUSE/DeepSea/issues/183
Signed-off-by: Tim Serong <tserong@suse.com>

tserong commented:
This is not ready, but I could use some feedback on a couple of things, notably:

1) Deploying mgrs is mandatory for Luminous+, and optional for Kraken. Does DeepSea know what version of ceph it's trying to deploy? Think: making the validate runner not require role-mgr if Jewel is being deployed. Or is that not a priority?

2) The auth key generation and everything here follows pretty much the same pattern as for MDS. This means DeepSea first generates keys in /srv/salt/ceph/mgr/cache, then later runs `ceph auth add`. If you already have a Kraken or Luminous cluster running (which right now will have the systemd magic auto-created ceph-mgr instances), *then* you add role-mgr to your policy.cfg, reflecting where the existing mgrs are, when you run stage 3 it will fail, because it's trying to replace existing mgr keys with newly created ones. Do we need to solve this?

tserong commented:
Actually, I can have it forcibly clobber any existing mgr keys with the ones DeepSea generates, which I think solves question 2 (see previous commit). Is this in any way bad?

To clarify the problem: in Kraken, and in current builds of Luminous, there's some systemd magic which causes mgr to start on all the MON nodes. If mgr keyrings don't exist, they will be generated automatically when the service starts, thanks to the magic. This means that if you add role-mgr to policy.cfg, whether or not you're running stage 3 to deploy a new cluster, or re-running it on an existing cluster, the ceph.mgr.auth piece will fail, because the keys DeepSea generates don't match the ones that already exist. Forcing replacement of the existing keys fixes this.

Once https://github.com/ceph/ceph/pull/16023 goes in, that problem evaporates for Luminous, so it won't cause trouble on new deployments, i.e. we don't need a forced overwrite in this case. But it still will be a problem for any existing Kraken or pre-release Luminous clusters managed by DeepSea.

I figure we can either go with my clobbering, or document that you should manually replace /srv/salt/ceph/mgr/cache/*keyring with keys from your existing cluster if you hit the error "Error EINVAL: entity mgr.HOSTNAME exists but key does not match" when running stage 3. Which of these two is least worst?

swiftgist commented:
Would it be possible to install a generated key for mgr prior to starting the mons?  That is, put the file in the right place during mon creation?

tserong commented:
Should be. That would fix it for new deploys, regardless of whether the systemd magic key generation is present or not (the magic only activates if there's not already a keyring). The problem would then only appear for existing Kraken or prerelease Luminous clusters.

tserong commented:
Yeah, seems to work, although for complete assurance I should really do a proper redeploy from scratch, rather than what I just did, which was manually rip the running mgrs out of an existing cluster then re-run stage 3.

Anyway, this means the problem will only remain for existing Kraken and prerelease Luminous clusters. In those cases, possibly documenting some manual screwing around is preferable to weird hacks?

swiftgist commented:
The current side effect is Stage 3 complaining that the keyring file on disk doesn't match the keyring in `ceph auth list`.  Asking the admin to choose between overwriting the cache file with the contents of `ceph auth get mgr.nodeX` or running `ceph auth del mgr.nodeX` and rerunning a Stage doesn't feel too terrible.

smithfarm commented:
> the problem will only remain for existing Kraken and prerelease Luminous clusters. In those cases, possibly documenting some manual screwing around is preferable to weird hacks

+1 

Supporting kraken and pre-release Luminous should be very low on the priority list IMO.

jschmid1 commented:
@smithfarm, @tserong, @swiftgist 
Just noticed that https://github.com/ceph/ceph/pull/16023 already found it's way into our codebase, which means that the default deployment is broken. 
We should merge the basic functionality asap and hand in remaining features as remove, rescind later.

smithfarm commented:
Yeah, sorry, I'm merging upstream master into the product on a regular basis - fail what may.

jschmid1 commented:
Not your fault @smithfarm :)
Just raising awareness that we should ship the corresponding fix asap

tserong commented:
@jschmid1, I've squashed that back to one commit so it can (hopefully) be merged immediately. I've also added mgr to the list in `srv/modules/runners/orderednodes.py:29`. I'm not sure about `srv/modules/runners/populate.py:519` -- there's a list on line 526, but I already added mgr to that earlier.

jschmid1 commented:
yikes, for the time being we need a `test.nop` in '/srv/salt/ceph/rescind/mgr' to avoid breakage

jschmid1 commented on srv/salt/ceph/stage/deploy/default.sls:
```diff
@@ -61,13 +61,32 @@ admin:
     - tgt_type: compound
     - sls: ceph.admin
 
+mgr auth:
```
Just had a timeout during deployment due to a, supposedly wrong ordering.

We try to auth against a monitor that is not yet deployed, see line 76.
This should be moved below the mon creation

tserong commented:
Please don't merge this yet, I might have missed something, need to test again.

tserong commented:
@jschmid1, the latest version implements rescind, and fixes the mgr auth ordering WRT mon creation.

jschmid1 commented:
giving it a spin @tserong

jschmid1 commented:
works like a charm! lgtm from my side
